src/main/java/org/breedinginsight/brapps/importer/services/processors/ExperimentProcessor.java

```diff
@@ -416,22 +462,44 @@ public void postBrapiData(Map<Integer, PendingImport> mappedBrAPIImport, Program
     private void prepareDataForValidation(List<BrAPIImport> importRows, List<Column<?>> phenotypeCols, Map<Integer, PendingImport> mappedBrAPIImport) {
         for (int rowNum = 0; rowNum < importRows.size(); rowNum++) {
             ExperimentObservation importRow = (ExperimentObservation) importRows.get(rowNum);
-
             PendingImport mappedImportRow = mappedBrAPIImport.getOrDefault(rowNum, new PendingImport());
-            mappedImportRow.setTrial(this.trialByNameNoScope.get(importRow.getExpTitle()));
-            mappedImportRow.setLocation(this.locationByName.get(importRow.getEnvLocation()));
-            mappedImportRow.setStudy(this.studyByNameNoScope.get(importRow.getEnv()));
-            mappedImportRow.setObservationUnit(this.observationUnitByNameNoScope.get(createObservationUnitKey(importRow)));
-
-            // loop over phenotype column observation data for current row
             List<PendingImportObject<BrAPIObservation>> observations = mappedImportRow.getObservations();
-            for (Column<?> column : phenotypeCols) {
-                // if value was blank won't be entry in map for this observation
-                observations.add(this.observationByHash.get(getImportObservationHash(importRow, getVariableNameFromColumn(column))));
-            }
+            String observationHash;
+            if (hasAllReferenceUnitIds) {
```

Comment:
I think in v0.10 we can try to make these workflows more modular and break things out a bit more rather than having if-else blocks in a number of different places